### PR TITLE
[UserList]: Return type for slice is set to whatever self is

### PR DESCRIPTION
**Pull request by raabf**
When the input of `__getitem__` is a slice, it returns an instance of the class whatever self currently is:

    def __getitem__(self, i):
        if isinstance(i, slice):
            return self.__class__(self.data[i])

https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/collections/__init__.py#L1215-L1217

That would be at least an instance of `UserList` (and not `MutableSequence` like currently stated in typeshed) but could be also a subclass. So I changed the return type to an instance of the class of `self`.

Side note: That problem is already addressed in the python2 typesheds:

https://github.com/python/typeshed/blob/f527e96dc30ed42a59f0a73ace2001561ce3b96c/stdlib/%40python2/UserList.pyi#L18

### CODE REVIEW

File: stdlib/collections/__init__.pyi [hunk 1]
```diff
@@ -69,7 +71,7 @@ class UserList(MutableSequence[_T]):
     @overload
     def __getitem__(self, i: int) -> _T: ...
     @overload
-    def __getitem__(self, i: slice) -> MutableSequence[_T]: ...
+    def __getitem__(self: _UserListT, i: slice) -> _UserListT[_T]: ...
```
Comment (srittau) [inline stdlib/collections/__init__.pyi:74]: We actually have a fairly new `Self` alias for that, which you can import from `_typeshed`:

```python
from _typeshed import Self

[...]

@overload
    def __getitem__(self: Self, i: slice) -> Self: ...
```

Comment (raabf) [reply stdlib/collections/__init__.pyi:74]: Oh that would be a very comfortable way. But does that work? `Self` from `_typeshed` is defined as `Self = TypeVar("Self")` (implicitly `bound=object`) and not `TypeVar("_UserListT", bound=UserList)` and the `UserList.__getitem__` implementation is accessing the attribute `self.data`, and the upper bound of `self` namely `object` does not have this attribute.

Comment (srittau) [reply stdlib/collections/__init__.pyi:74]: A typevar annotating `self` is implicitly bound to that class. I'm not sure this is documented anywhere, but this is how type checkers work.

Comment (raabf) [reply stdlib/collections/__init__.pyi]: At least I never read that, and hence did not know about that, … but that behaviour makes life easier. Thanks for explanation! I changed the PR to `_typesheld.Self`.

**Review by srittau — APPROVED** [review]

**Pull request merged**